sphinx/domains/python.py

```diff
@@ -1159,8 +1202,14 @@ def resolve_xref(self, env: BuildEnvironment, fromdocname: str, builder: Builder
                                 type, searchmode)
 
         if not matches and type == 'attr':
-            # fallback to meth (for property)
+            # fallback to prop (for property)
+            matches = self.find_obj(env, modname, clsname, target, 'prop', searchmode)
+        if not matches and type == 'attr':
+            # fallback to meth (for property; Sphinx-2.4.x)
             matches = self.find_obj(env, modname, clsname, target, 'meth', searchmode)
+        if not matches and type == 'meth':
+            # fallback to prop (for property)
+            matches = self.find_obj(env, modname, clsname, target, 'prop', searchmode)
```

Comment:
You also need to add the following, right?

Since people can transparently change attributes to properties, too?

```py
if not matches and type == 'prop':
    matches = self.find_obj(env, modname, clsname, target, 'attr', searchmode)
```

And maybe more informative comments would be:

1. for attr/prop:
   
   ```py
   # properties and attributes are transparently interchangeable,
   # so their roles need to be able to refer to the respective other
   ```

2. for `meth`:

   ```py
   # fallback to meth (in Sphinx 2.4.x, properties were stored as py:method)
   ```

Reply:
Never discussed that `py:prop` role can refer to attributes. And I don't think so.

Reply:
It should, as people can always also transparently change properties to attributes. It’s going to happen rarely, but it will.

Reply:
I haven't followed the full discussion here, but in my opinion either:

* Only `:attr:` should exist
* `:attr:` / `:prop:` should be completely interchangeable

I'd argue that while these's some value to describing `.. attribute` and `.. property` differently 
 to draw attention to dynamic behavior, there's no value in referencing them differently, as to a consumer there's no difference.

Reply:
That’s been my argument all along. @tk0miya thinks the consistency of having a one role per directive is important, while I tended towards going more in the “Only `:attr:` should exist” direction, because of the Python zen: “there should be one – and preferably only one – obvious way to do it”. But @tk0miya and me had a loooong discussion that we don’t want to repeat so I think the best compromise is “`:attr:` / `:prop:` should be completely interchangeable”

Reply:
@eric-wieser Who is the "producer" (provider?) and "consumer"?

I think no reason to describe them differently if they are transparent. Is it worthy to describe them differently? If so, they're different things, I think.

Reply:
To be clear, the "provider" is the docs that use `.. py:property`, such as a framework that provides the property. A "consumer" is a docstring in a docstream package that wants to reference the property.

> I think no reason to describe them differently if they are transparent. Is it worthy to describe them differently? If so, they're different things, I think.

Let's use `:abstract:` as an example - it's useful for readers to know that a method is abstract when reading its docstring, but when linking `` :abstract-meth:`some_meth` `` wouldn't be useful.

So by that reasoning I'd argue that `:property:` and `:abstract:` should be treated the same way
```
.. attribute:: my_prop
    :property:

    This is the doc for :attr:`my_prop`
```
I realize I'm at risk of reviving @flying-sheep's discussion here though.

Reply:
Do you mean to use `py:attribute` directive to describe property (as an option)? If so, I can understand your idea. But I choose to add a new type `py:property` because it is unnatural to describe "abstract" or "async" properties via `py:attribute` directive. Additionally, client library sometimes needs to know the "member" of object is an attribute or a property (ex. assigning value, inheritance, etc).

Should we discuss "what is a property"?

Reply:
> Do you mean to use py:attribute directive to describe property (as an option)

Yes, I guess I was suggesting that

> because it is unnatural to describe "abstract" or "async" properties

The former is `@abstractproperty`, and the latter has a PyPI package; https://pypi.org/project/async-property/.

So I think actually that's an argument _for_ `:property:` just being metadata used in `..attribute`